### ISSUE TITLE
Temporary file is deleted before a controller returns a response

### ISSUE THREAD

**Issue by tkawachi** [playframework/playframework#12770]
### Play Version

3.0.4

### API

Scala

### Operating System

macOS 14.4.1

### JDK

```
openjdk version "21.0.2" 2024-01-16 LTS
OpenJDK Runtime Environment Corretto-21.0.2.13.1 (build 21.0.2+13-LTS)
OpenJDK 64-Bit Server VM Corretto-21.0.2.13.1 (build 21.0.2+13-LTS, mixed mode, sharing)
```

### Library Dependencies

None

### Expected Behavior

I have prepared the following Controller method.

```scala
  def fileSize() =
    Action.async(parse.multipartFormData) { implicit request =>
      request.body.file("file") match {
        case Some(image) =>
          for {
            path <- Future(image.ref.path)
            _ <- Future(System.gc())
            _ <- Future(Thread.sleep(100))
          } yield Ok(Files.size(path).toString)
        case None =>
          Future.successful(BadRequest("No image supplied"))
      }
    }
```

The endpoint is as follows.

```
POST    /fileSize                   controllers.HomeController.fileSize()
```

1. Send a POST request.
    - `echo -n Hello > /tmp/a; curl -X POST http://localhost:9000/fileSize -F"file=@/tmp/a"`
2. The server returns the file size and the command displays 5.

### Actual Behavior

1. Send a POST request.
    - `echo -n Hello > /tmp/a; curl -X POST http://localhost:9000/fileSize -F"file=@/tmp/a"`
2. The server returns an error page.
3. The server logs following

```
2024-06-18 18:25:01 ERROR p.api.http.DefaultHttpErrorHandler  

! @85ph4o99f - Internal server error, for (POST) [/fileSize] ->
 
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[NoSuchFileException: /var/folders/r2/qjlxkmsn49v6mtmb6dgs3xrc0000gr/T/playtemp8649900777384842915/multipartBody4548848415921731547asTemporaryFile]]
	at play.api.http.HttpErrorHandlerExceptions$.$anonfun$convertToPlayException$2(HttpErrorHandler.scala:400)
	at scala.Option.map(Option.scala:242)
	at play.api.http.HttpErrorHandlerExceptions$.convertToPlayException(HttpErrorHandler.scala:398)
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:390)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:267)
	at play.core.server.PekkoHttpServer$$anonfun$invokeAction$1$1.applyOrElse(PekkoHttpServer.scala:482)
	at play.core.server.PekkoHttpServer$$anonfun$invokeAction$1$1.applyOrElse(PekkoHttpServer.scala:474)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:490)
	at org.apache.pekko.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:73)
	at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:110)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
	at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:110)
	at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
	at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.nio.file.NoSuchFileException: /var/folders/r2/qjlxkmsn49v6mtmb6dgs3xrc0000gr/T/playtemp8649900777384842915/multipartBody4548848415921731547asTemporaryFile
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:171)
	at java.base/java.nio.file.Files.readAttributes(Files.java:1853)
	at java.base/java.nio.file.Files.size(Files.java:2462)
	at controllers.HomeController.$anonfun$fileSize$7(HomeController.scala:36)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	... 5 common frames omitted
```

The file pointed to by TemporaryFile seems to be erased during GC, but I suppose that it would not be GC'd until it returns a response.

### Reproducible Test Case

https://github.com/tkawachi/play-tmpfile-deleted/commit/c9dbcd7c00863afc5b2f070cdc7fe9ceabe85dda

**Comment by mkurz** [comment]
Actually this code is working as intended. The "problem" you run into is that after retrieving the `path` of `image.ref` you do not access `image.ref` anymore, so there is no need for the JVM to keep a reference to the object if points to and "releases" the object it points to for garbage collection. When the object gets "destroyed" it will also remove its temporary file, you can see that here:
https://github.com/playframework/playframework/blob/814d1921b1f8eb7a743db5f93178b49b26fd8232/core/play/src/main/scala/play/api/libs/Files.scala#L281-L286
Because you force garbage collection and afterwards access the `path`, the file `path` represents is already deleted.
I see you added a `sleep` statement to make sure Garbage collection finished before you call `Files.size` on the temp file. Without the `sleep` statement sometimes the code works, sometimes it fails. Depending if Garbage collection is fast enough to delete the file or not.
Now, for me this is not a bug, but just how this is designed to work. So you have to understand as soon you do not reference `image.ref` anymore you can not be sure that the file it represents on the file system still exists afterwards. So IMHO you have two choices:
1. Either move the temporary file to a different location so it will not be deleted when the `image.ref` gets destroyed and then only work with that "new" file, e.g.:
    ```scala
        case Some(image) =>
          for {
            path <- Future(image.ref.atomicMoveWithFallback(someTargetPath))
            ...
    ```
    Be aware that `someTargetPath`/`path` will not be deleted automatically anymore, it's managed by you and you have to make sure to remove it or do whatever you want with it.
1. Do not save a reference to `image.ref.path`. If you do not access `image.ref` anymore it means the file it references is never safe to access. In your code you could do following instead:
    ```scala
          for {
            fsize <- Future(Files.size(image.ref.path))
            _ <- Future(System.gc())
            _ <- Future(Thread.sleep(100))
          } yield Ok(fsize.toString)
    ```

- [ ] We might want to be more clear about how things work in the the documentation:
https://www.playframework.com/documentation/3.0.x/ScalaFileUpload#Cleaning-up-temporary-files

- [ ] Also actually your example makes a good (scripted?) test to make sure that temporary files get deleted when the `ref` object gets destroyed (not sure how intense we test this yet).

**Comment by tkawachi** [comment]
Thank you for your response.

I understand the mechanism, but I feel that not retaining the reference to `image.ref.path` in 2. makes it prone to programming errors and difficult to handle.

As a third option, it seems possible to retain the reference to the request until the end. By retaining the reference to the request, the reference to the TemporaryFile is also retained, so the file will be maintained.

```scala
  def fileSize() =
    Action.async(parse.multipartFormData) { implicit request =>
      request.body.file("file") match {
        case Some(image) =>
          for {
            path <- Future(image.ref.path)
            _ <- Future(System.gc())
            _ <- Future(Thread.sleep(100))
          } yield Ok{
            request // just for retaining a reference
            Files.size(path).toString
          }
        case None =>
          Future.successful(BadRequest("No image supplied"))
      }
    }
```

Is it possible to implement this on the framework side?

Regarding 1., I am a bit concerned about the extra I/O that occurs.

**Comment by mkurz** [comment]
That's an interesting idea. Actually yesterday I tried
```scala
...
            _ <- Future(Thread.sleep(100))
            _ <- Future(image.ref) // <--
          } yield Ok(Files.size(path).toString)
...
```
to keep the object alive, which also worked, but haven't thought about to keep a reference to request within Play itself.
Your solution makes sense. Could you open a pull request? Would be nice if you could add a test also. fyi, afaik `System.gc()` does not guarantee that GC will actually be triggered, it just asks the JVM to do so. So _in theory_ such a test is not 100% reliable, but I guess since we can both reproduce the exception, you with corretto on macOS and I with Temurin on Linux, _in practise_ we can say it's safe to assume `System.gc()` does what we want for such a test.

**Comment by mkurz** [comment]
@tkawachi Play 3.0.5 with the fixed was just released.

**Comment by weichengliao11** [comment]
Hi @mkurz 
I guess I encountered the same issue on Play 2.6.1, so I'm curious if you fixed it in the previous version or I need to find another way out?

**Comment by mkurz** [comment]
@weichengliao11 This is only fixed in latest Play 2.9.x and 3.0.x releases. All previous Play releases are not supported anymore.
